src/sentry/services/hybrid_cloud/organization_mapping/impl.py

```diff
@@ -0,0 +1,75 @@
+from dataclasses import fields
+from typing import Any, Optional
+
+from django.db import IntegrityError, transaction
+
+from sentry.models.organizationmapping import OrganizationMapping
+from sentry.models.user import User
+from sentry.services.hybrid_cloud.organization_mapping import (
+    APIOrganizationMapping,
+    OrganizationMappingService,
+)
+
+
+class DatabaseBackedOrganizationMappingService(OrganizationMappingService):
+    def create(
+        self,
+        *,
+        user: User,
+        organization_id: int,
+        slug: str,
+        region_name: str,
+        idempotency_key: Optional[str] = "",
+        # There's only a customer_id when updating an org slug
+        customer_id: Optional[str] = None,
+    ) -> APIOrganizationMapping:
+        try:
+            with transaction.atomic():
+                # Creating an identical mapping should succeed, even if a record already exists
+                # with this slug. We allow this IFF the idempotency key is identical.
+                org_mapping = OrganizationMapping.objects.create(
+                    organization_id=organization_id,
+                    slug=slug,
+                    customer_id=customer_id,
+                    idempotency_key=idempotency_key,
+                    region_name=region_name,
+                )
+            return self.serialize_organization_mapping(org_mapping)
+        except IntegrityError:
+            pass
+
+        # If we got here, the slug already exists
+        if idempotency_key != "":
+            try:
+                with transaction.atomic():
+                    existing_mapping = OrganizationMapping.objects.select_for_update().get(
```

Comment:
As crazy as it sounds, it is theoretically possible that the existing_mapping here is None in the case of a concurrent delete.  Could we raise some error message indicating a concurrent transaction?  Just so that it's not super confusing to someone.  This error condition is expected and not indicative of a bug but just general concurrency issues.

Reply:
@corps I moved this to an update_or_create to better handle concurrent deletions. Right now, I'm not sure we can tell the difference between a concurrent transaction and an organization slug already existing.